### PR TITLE
Rethink api

### PR DESCRIPTION
**Pull request by lukasbuenger**
Please review as you see fit.

Notable stuff:

* [BasicDocument](https://github.com/orbiting/haku-frontend/compare/rethinkApi?expand=1#diff-0f75b421da9299aa8ec12a31055e58f9) - gives you a feel of how a document gets composed.
* [Editor](https://github.com/orbiting/haku-frontend/compare/rethinkApi?expand=1#diff-2fe74c384fab281b986b003b54d36b9e) - here you see how such a document gets selected, rendered and controlled.
* The link module ([files below here](https://github.com/orbiting/haku-frontend/compare/rethinkApi?expand=1#diff-eb165f3b2830df3245cf9f2eb31251d6))  - has a property form

The utils folder contains a lot of very basic stuff, that's why I decided to have everything in there at least somehow tested. It's a start I guess. I'd recommend:

* [createFormatButton](https://github.com/orbiting/haku-frontend/compare/rethinkApi?expand=1#diff-9b24d0e4c37dc96782647063585fa84a) - Very basic HOC 
* [createMarkButton](https://github.com/orbiting/haku-frontend/compare/rethinkApi?expand=1#diff-f6820459d32bb1177e7599d7e12cba45) - Decorates the former
* [Scary test shit](https://github.com/orbiting/haku-frontend/compare/rethinkApi?expand=1#diff-ca0625b182fe77d07d6c71c8797a05d3) - Because that's what it is.

The rule composition thing is already working, but naming is still meh. Bear with me.

Can get imported from `components/editor/Editor` and used as drop-in-replacement of the previous import.  

That's it for now.

### CODE REVIEW

File: components/editor/Editor.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+  render () {
+    const { state } = this.props
+    const { lockedState } = this.state
+    const UI = getUI(state)
+    const props = {
+      state: lockedState || state,
+      onChange: this.changeHandler,
+      onClaimLock: this.claimLockHander,
+      onReleaseLock: this.releaseLockHander,
```
Comment (tpreusse) [inline components/editor/Editor.js]: what are those locks for? seem unused. Also should probably not be named `onClaimLock/onRelease` here as it's an action the document component triggers and not an event.

File: components/editor/BasicDocument.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+              key={`form-${i}`}
+              state={state}
+              onChange={onChange}
+              />
+          ))
+        }
+      </Interaction.P>
+    </Sidebar>
+    <Document>
+      <NarrowContainer>
```
Comment (tpreusse) [inline components/editor/BasicDocument.js]: Narrow container should probably be part of the document passed in, a bit confusing to have a dynamic configurable container and a hard coded one

File: components/editor/modules/lead/ui.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+      {...{...css(styles.blockButton), ...props}}
+      data-active={active}
+      data-disabled={disabled}
+      >
+      Lead
+    </span>
+)
+
+export default {
+  LeadButton
```
Comment (tpreusse) [inline components/editor/modules/lead/ui.js]: why double export—named and default object with the same name?

File: components/editor/utils/createMarkButton.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        const start = i
+        i = offset
+        has = true
+        while (has) {
+          i++
+          has = characters.get(i).marks.some(matchMark(markType))
+        }
+        const end = i
+        const length = end - start
+        return onChange && onChange(
```
Comment (tpreusse) [inline components/editor/utils/createMarkButton.js]: in which case can `onChange` be missing and why?

Comment (tpreusse) [inline components/editor/utils/createMarkButton.js]: maybe a default prop doing noop instead?

File: components/editor/utils/index.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+export const matchBlock = match('block')
+
+export const matchMark = match('mark')
+
+export const matchInline = match('inline')
+
+export const matchDocument = node =>
+  node.kind === 'document'
+
+export const pluginFromRules = rules =>
```
Comment (tpreusse) [inline components/editor/utils/index.js]: I was scared of a smart factory doing too much and then disappointed by this. Really necessary to abstract this?

**Comment by tpreusse** [comment]
Missing: Image

File: components/editor/BasicDocument.js (hunk 2 above)
Comment (lukasbuenger) [inline components/editor/BasicDocument.js]: As a matter of fact it is. In this case `NarrowContainer` only wraps the slate `Editor` if the passed document is of type `BasicDocument` (which as for know is always). Slate itself does not control the way the document aka the root node is rendered and this, to me at least, seemed like the most unobtrusive way to do it.

**Pull request merged**